engine/src/test/java/org/hibernate/search/test/projection/ProjectionConversionTest.java

```diff
@@ -0,0 +1,241 @@
+import org.hibernate.search.query.engine.spi.EntityInfo;
+import org.hibernate.search.testsupport.TestForIssue;
+import org.hibernate.search.testsupport.concurrency.ConcurrentRunner;
+import org.hibernate.search.testsupport.concurrency.ConcurrentRunner.TaskFactory;
+import org.hibernate.search.testsupport.junit.SearchFactoryHolder;
+import org.hibernate.search.testsupport.setup.TransactionContextForTest;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * This test verifies the correct projection rules (like which FieldBridge)
+ * needs to be applied on each projected field.
+ * As witnessed by HSEARCH-1786 and HSEARCH-1814, it is possible that when
+ * having multiple entity types or non trivial entity graphs that the same
+ * field name has different mapping rules per type, and conflicts are possible
+ * when the projection handling code is not careful about this.
+ *
+ * @author Sanne Grinovero <sanne@hibernate.org> (C) 2015 Red Hat Inc.
+ */
+@TestForIssue(jiraKey = "HSEARCH-1786")
+public class ProjectionConversionTest {
+
+	@Rule
+	public SearchFactoryHolder sfHolder = new SearchFactoryHolder( ExampleEntity.class );
+
+	@Before
+	public void storeTestData() {
+		ExtendedSearchIntegrator searchFactory = sfHolder.getSearchFactory();
+		ExampleEntity entity = new ExampleEntity();
+		entity.id = 1l;
+		entity.someInteger = 5;
+		entity.longEncodedAsText = 20l;
+		entity.customBridgedKeyword = "lowercase-keyword";
+
+		ExampleEntity embedded = new ExampleEntity();
+		embedded.id = 2l;
+		embedded.someInteger = 6;
+		embedded.longEncodedAsText = 21l;
+		embedded.customBridgedKeyword = "another-lowercase-keyword";
+
+		ConflictingMappedType second = new ConflictingMappedType();
+		second.id = "a string";
+		second.customBridgedKeyword = 17l;
+
+		entity.embedded = embedded;
+		entity.second = second;
+
+		Work work = new Work( entity, entity.id, WorkType.ADD, false );
+		TransactionContextForTest tc = new TransactionContextForTest();
+		searchFactory.getWorker().performWork( work, tc );
+		tc.end();
+	}
+
+	@Test
+	public void projectingExplicitId() {
+		projectionTestHelper( ProjectionConstants.ID, Long.valueOf( 1l ) );
+	}
+
+	@Test
+	public void projectingIdByPropertyName() {
+		projectionTestHelper( "id", Long.valueOf( 1l ) );
+	}
+
+	@Test
+	public void projectingIdOnOverloadedMapping() {
+		projectionTestHelper( "stringTypedId", Long.valueOf( 1l ) );
+	}
+
+	@Test
+	public void projectingIntegerField() {
+		projectionTestHelper( "someInteger", Integer.valueOf( 5 ) );
+	}
+
+	@Test
+	public void projectingUnknownField() {
+		projectionTestHelper( "someNonExistingField", null );
+	}
+
+	@Test
+	public void projectionWithCustomBridge() {
+		projectionTestHelper( "customBridgedKeyword", "lowercase-keyword" );
+	}
+
+	@Test
+	public void projectingEmbeddedIdByPropertyName() {
+		projectionTestHelper( "embedded.id", Long.valueOf( 2l ) ); //FAILS OFTEN
```

Comment:
I'll remove it :)
It's a left over from notes on debugging. It was very puzzling to figure out, so I had some notes and some ugly system-outs in various places

thanks, good catch!